File: LoginFrame.java
import javax.swing.*;
import java.awt.*;
public class LoginFrame extends JFrame {
    private static final long serialVersionUID = 1L;
    private final DatabaseManager dbManager;
    private JTextField usernameField;
    private JPasswordField passwordField;
    private JTextField nameField;
    private JTextField emailField;
    private User loggedInUser;
    public LoginFrame() {
        dbManager = new DatabaseManager();
        loggedInUser = null;
        setTitle("Habit Tracker - Login");
        setSize(400, 300);
        setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        setLocationRelativeTo(null);
        setLayout(new BorderLayout(10, 10));
        JTabbedPane tabbedPane = new JTabbedPane();
        tabbedPane.addTab("Login", createLoginPanel());
        tabbedPane.addTab("Register", createRegisterPanel());
        add(tabbedPane, BorderLayout.CENTER);
        setVisible(true);
    }
    private JPanel createLoginPanel() {
        JPanel panel = new JPanel(new GridBagLayout());
        panel.setBorder(BorderFactory.createEmptyBorder(20, 20, 20, 20));
        GridBagConstraints gbc = new GridBagConstraints();
        gbc.insets = new Insets(5, 5, 5, 5);
        gbc.fill = GridBagConstraints.HORIZONTAL;
        JLabel titleLabel = new JLabel("Login to Habit Tracker");
        titleLabel.setFont(new Font("Arial", Font.BOLD, 16));
        gbc.gridx = 0;
        gbc.gridy = 0;
        gbc.gridwidth = 2;
        gbc.anchor = GridBagConstraints.CENTER;
        panel.add(titleLabel, gbc);
        gbc.gridwidth = 1;
        gbc.gridy = 1;
        gbc.anchor = GridBagConstraints.WEST;
        panel.add(new JLabel("Username:"), gbc);
        usernameField = new JTextField(20);
        gbc.gridx = 1;
        panel.add(usernameField, gbc);
        gbc.gridx = 0;
        gbc.gridy = 2;
        panel.add(new JLabel("Password:"), gbc);
        passwordField = new JPasswordField(20);
        gbc.gridx = 1;
        panel.add(passwordField, gbc);
        JButton loginButton = new JButton("Login");
        loginButton.addActionListener(e -> handleLogin());
        gbc.gridx = 0;
        gbc.gridy = 3;
        gbc.gridwidth = 2;
        gbc.anchor = GridBagConstraints.CENTER;
        panel.add(loginButton, gbc);
        passwordField.addActionListener(e -> handleLogin());
        return panel;
    }
    private JPanel createRegisterPanel() {
        JPanel panel = new JPanel(new GridBagLayout());
        panel.setBorder(BorderFactory.createEmptyBorder(20, 20, 20, 20));
        GridBagConstraints gbc = new GridBagConstraints();
        gbc.insets = new Insets(5, 5, 5, 5);
        gbc.fill = GridBagConstraints.HORIZONTAL;
        JLabel titleLabel = new JLabel("Create New Account");
        titleLabel.setFont(new Font("Arial", Font.BOLD, 16));
        gbc.gridx = 0;
        gbc.gridy = 0;
        gbc.gridwidth = 2;
        gbc.anchor = GridBagConstraints.CENTER;
        panel.add(titleLabel, gbc);
        gbc.gridwidth = 1;
        gbc.gridy = 1;
        gbc.anchor = GridBagConstraints.WEST;
        panel.add(new JLabel("Full Name:"), gbc);
        nameField = new JTextField(20);
        gbc.gridx = 1;
        panel.add(nameField, gbc);
        gbc.gridx = 0;
        gbc.gridy = 2;
        panel.add(new JLabel("Email:"), gbc);
        emailField = new JTextField(20);
        gbc.gridx = 1;
        panel.add(emailField, gbc);
        gbc.gridx = 0;
        gbc.gridy = 3;
        panel.add(new JLabel("Username:"), gbc);
        JTextField regUsernameField = new JTextField(20);
        gbc.gridx = 1;
        panel.add(regUsernameField, gbc);
        gbc.gridx = 0;
        gbc.gridy = 4;
        panel.add(new JLabel("Password:"), gbc);
        JPasswordField regPasswordField = new JPasswordField(20);
        gbc.gridx = 1;
        panel.add(regPasswordField, gbc);
        JButton registerButton = new JButton("Create Account");
        registerButton.addActionListener(e -> handleRegister(
            nameField.getText(),
            emailField.getText(),
            regUsernameField.getText(),
            new String(regPasswordField.getPassword())
        ));
        gbc.gridx = 0;
        gbc.gridy = 5;
        gbc.gridwidth = 2;
        gbc.anchor = GridBagConstraints.CENTER;
        panel.add(registerButton, gbc);
        return panel;
    }
    private void handleLogin() {
        String username = usernameField.getText().trim();
        String password = new String(passwordField.getPassword());
        if (username.isEmpty() || password.isEmpty()) {
            JOptionPane.showMessageDialog(this, 
                "Please enter both username and password!", 
                "Login Error", 
                JOptionPane.ERROR_MESSAGE);
            return;
        }
        loggedInUser = dbManager.loginUser(username, password);
        if (loggedInUser != null) {
            JOptionPane.showMessageDialog(this, 
                "Welcome back, " + loggedInUser.getName() + "!", 
                "Login Successful", 
                JOptionPane.INFORMATION_MESSAGE);
            openMainApplication();
        } else {
            JOptionPane.showMessageDialog(this, 
                "Invalid username or password!", 
                "Login Failed", 
                JOptionPane.ERROR_MESSAGE);
            passwordField.setText("");
        }
    }
    private void handleRegister(String name, String email, String username, String password) {
        if (name.isEmpty() || email.isEmpty() || username.isEmpty() || password.isEmpty()) {
            JOptionPane.showMessageDialog(this, 
                "Please fill in all fields!", 
                "Registration Error", 
                JOptionPane.ERROR_MESSAGE);
            return;
        }
        if (password.length() < 6) {
            JOptionPane.showMessageDialog(this, 
                "Password must be at least 6 characters long!", 
                "Registration Error", 
                JOptionPane.ERROR_MESSAGE);
            return;
        }
        int userId = dbManager.registerUser(name, username, email, password);
        if (userId > 0) {
            JOptionPane.showMessageDialog(this, 
                "Account created successfully! Please login.", 
                "Registration Successful", 
                JOptionPane.INFORMATION_MESSAGE);
            nameField.setText("");
            emailField.setText("");
        } else {
            JOptionPane.showMessageDialog(this, 
                "Registration failed! Username may already exist.", 
                "Registration Error", 
                JOptionPane.ERROR_MESSAGE);
        }
    }
    private void openMainApplication() {
        SwingUtilities.invokeLater(() -> {
            new HabitTrackerApp(loggedInUser, dbManager);
            dispose();
        });
    }
    public static void main(String[] args) {
        try {
            UIManager.setLookAndFeel(UIManager.getSystemLookAndFeelClassName());
        } catch (Exception e) {
            e.printStackTrace();
        }
        SwingUtilities.invokeLater(() -> {
            new LoginFrame();
        });
    }
}

File: LoginFrame.py
import tkinter as tk
from tkinter import messagebox

class DatabaseManager:
    def __init__(self):
        pass

    def login_user(self, username, password):
        # implement database logic to login a user
        pass

    def register_user(self, name, username, email, password):
        # implement database logic to register a user
        pass

class HabitTrackerApp:
    def __init__(self, user, db_manager):
        # implement the habit tracker application
        pass

class LoginFrame:
    def __init__(self):
        self.db_manager = DatabaseManager()
        self.logged_in_user = None
        self.root = tk.Tk()
        self.root.title("Habit Tracker - Login")
        self.root.geometry("400x300")
        self.root.config(padx=10, pady=10)

        self.tab_control = tk.Notebook(self.root)
        self.login_tab = tk.Frame(self.tab_control)
        self.register_tab = tk.Frame(self.tab_control)
        self.tab_control.add(self.login_tab, text="Login")
        self.tab_control.add(self.register_tab, text="Register")
        self.tab_control.pack(expand=1, fill="both")

        self.create_login_tab()
        self.create_register_tab()

        self.root.mainloop()

    def create_login_tab(self):
        login_frame = tk.Frame(self.login_tab)
        login_frame.pack(fill="both", expand=1)

        tk.Label(login_frame, text="Login to Habit Tracker", font=("Arial", 16, "bold")).grid(row=0, column=0, columnspan=2, padx=10, pady=10)

        tk.Label(login_frame, text="Username:").grid(row=1, column=0, padx=5, pady=5)
        self.username_field = tk.Entry(login_frame)
        self.username_field.grid(row=1, column=1, padx=5, pady=5)

        tk.Label(login_frame, text="Password:").grid(row=2, column=0, padx=5, pady=5)
        self.password_field = tk.Entry(login_frame, show="*")
        self.password_field.grid(row=2, column=1, padx=5, pady=5)

        login_button = tk.Button(login_frame, text="Login", command=self.handle_login)
        login_button.grid(row=3, column=0, columnspan=2, padx=10, pady=10)

    def create_register_tab(self):
        register_frame = tk.Frame(self.register_tab)
        register_frame.pack(fill="both", expand=1)

        tk.Label(register_frame, text="Create New Account", font=("Arial", 16, "bold")).grid(row=0, column=0, columnspan=2, padx=10, pady=10)

        tk.Label(register_frame, text="Full Name:").grid(row=1, column=0, padx=5, pady=5)
        self.name_field = tk.Entry(register_frame)
        self.name_field.grid(row=1, column=1, padx=5, pady=5)

        tk.Label(register_frame, text="Email:").grid(row=2, column=0, padx=5, pady=5)
        self.email_field = tk.Entry(register_frame)
        self.email_field.grid(row=2, column=1, padx=5, pady=5)

        tk.Label(register_frame, text="Username:").grid(row=3, column=0, padx=5, pady=5)
        self.reg_username_field = tk.Entry(register_frame)
        self.reg_username_field.grid(row=3, column=1, padx=5, pady=5)

        tk.Label(register_frame, text="Password:").grid(row=4, column=0, padx=5, pady=5)
        self.reg_password_field = tk.Entry(register_frame, show="*")
        self.reg_password_field.grid(row=4, column=1, padx=5, pady=5)

        register_button = tk.Button(register_frame, text="Create Account", command=self.handle_register)
        register_button.grid(row=5, column=0, columnspan=2, padx=10, pady=10)

    def handle_login(self):
        username = self.username_field.get()
        password = self.password_field.get()

        if not username or not password:
            messagebox.showerror("Login Error", "Please enter both username and password!")
            return

        self.logged_in_user = self.db_manager.login_user(username, password)

        if self.logged_in_user:
            messagebox.showinfo("Login Successful", f"Welcome back, {self.logged_in_user}!")
            self.open_main_application()
        else:
            messagebox.showerror("Login Failed", "Invalid username or password!")
            self.password_field.delete(0, tk.END)

    def handle_register(self):
        name = self.name_field.get()
        email = self.email_field.get()
        username = self.reg_username_field.get()
        password = self.reg_password_field.get()

        if not name or not email or not username or not password:
            messagebox.showerror("Registration Error", "Please fill in all fields!")
            return

        if len(password) < 6:
            messagebox.showerror("Registration Error", "Password must be at least 6 characters long!")
            return

        user_id = self.db_manager.register_user(name, username, email, password)

        if user_id > 0:
            messagebox.showinfo("Registration Successful", "Account created successfully! Please login.")
            self.name_field.delete(0, tk.END)
            self.email_field.delete(0, tk.END)
        else:
            messagebox.showerror("Registration Error", "Registration failed! Username may already exist.")

    def open_main_application(self):
        self.root.destroy()
        HabitTrackerApp(self.logged_in_user, self.db_manager)

if __name__ == "__main__":
    LoginFrame()